### PR TITLE
Security: Avoid reconnecting to peers that are likely unreachable

### LINKED ISSUE #1865: Security: Running Zebra nodes should eventually stop trying to contact peers that always fail
## Motivation

Zebra will keep trying individual `Failed` peers, even if they have never succeeded. This is a distributed denial of service risk, and it places extra load on the network.

## Scheduling

We should fix this issue before NU5 mainnet activation, so this bug doesn't cause a denial of service from old Zebra versions when _NU6_ activates.

## Suggestions

This fix depends on #1849, #1867, and #1871.

Zebra should stop trying to contact peers that haven't had a successful connection for 3 days. We've chosen this time to allow admins to restart their nodes after a weekend failure. (We might want to change this to a longer timeframe in a future upgrade, once Zebra is stable.)

Zebra should delete peers from the AddressBook where the:
- [ ] `last_success_time` is older than 3 days
- [ ] `last_success_time` is `None`, and the `untrusted_last_seen` is older than 3 days (requires the far-future fix in #1871)
- [ ] this task should be triggered by an interval timer that runs every minute, see `crawl_and_dial` for an example of this kind of address book task
  - the task should directly run an address book method containing these checks, to make testing easier
  - the task should also be run if the address book gets too large, but we'll do that in #1873

Zebra should also:
- [x] provide an accessor method on `MetaAddr` to simplify the interface to these times

To avoid sending old peers to other nodes, Zebra should:
- [x] keep the earliest last seen times gossiped by other peers
- [x] avoid sending unreachable DNS seeder addresses to peers
  - these addresses don't have any times, so we have no way of removing them from the network
- [ ] before sending unreachable gossiped addresses, subtract 30 minutes from their last seen times
  -  this effectively becomes a time-to-live for unreachable addresses, giving them 6 gossips before Zebra stops forwarding them (based on the 3 hour gossip limit from #1867)
  - the Bitcoin reference says we shouldn't change these times, but we already sanitize them

To avoid accepting old peers from other nodes, Zebra should:
- [ ]  ignore peers that are older than 3 days
  - Zebra should count back 3 days from the newest peer timestamp sent by the other peer, to compensate for clock skew (otherwise, different local and remote clocks could make us delete all the peers)
  - For details, see the clock skew adjustment algorithm in #1871

Property testing:
- [ ] Generate some random `MetaAddr`s
- [ ] Put them in an `AddressBook`
- [ ] Run the deletion method on the address book
- [ ] Make sure that old `last_success_time`s and `untrusted_last_seen_time`s are handled correctly

For testing:
- [ ] add a `debug_peer_deletion_age` config that sets the deletion age **and** interval timer
- [ ] add an acceptance test that sets it to `15` seconds
- [ ] add a log when all the peers are deleted, and check for that log - "hint: upgrade or restart Zebra"

## Performance Analysis

If we check for old peers every time a new peer is requested, we could spend a lot of time checking for deletions. Instead, we should scan the address book at regular intervals in a new task.

## Alternatives

This is a critical security issue, so we must do something.

We could keep a failure count for each peer. This design has usability issues on unreliable networks, because all the peers can fail at the same time. (Our reconnection rate limit and peer deletion timeout already limit us to ~2000 failed connections per peer over 3 days.)

We could filter peers whenever the `AddressBook` is accessed, rather than deleting them. But this is tricky - it might need interior mutability. It would also require an abstraction layer, to make sure we intercept all accesses.

We could avoid deleting peers in `NeverAttempted...` states, so that we try each peer at least once before deleting it. But this would risk a memory DoS, if we get a lot of gossiped peers.

## Context

`zcashd` does not have this issue.

`zcashd` has a support interval of around 16 weeks between required upgrades, with new versions coming out every 6 weeks.

## Follow-Up Tasks

This fix doesn't deal with denial of service from nodes that are regularly restarted, that's #1870.

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

Zebra would keep trying to connect to `Failed` peers, even if they are likely unreachable and will always fail. This is a Denial-of-Service risk and places extra load on the network.

This closes #1865.

## Solution

Filter the list of peers to reconnect to so that it doesn't contain peers that are likely unreachable.

A peer is considered to be probably unreachable if it was last seen (or was reported last seen) more than three days ago and if the previous connection attempt to it failed.

## Review

I had discussed the implementation solution with @teor2345.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

Peers that are likely to be unreachable should be removed from the `AddressBook`. This might be part of #1873.

### CODE REVIEW

File: zebra-network/src/meta_addr.rs [hunk 1]
```diff
@@ -521,6 +521,24 @@ impl MetaAddr {
...
+
+    /// Was this peer last seen a long time ago?
+    ///
+    /// A long time in this case is more than 3 days ago or never seen before.
+    pub fn was_not_recently_seen(&self) -> bool {
+        match self.last_seen() {
+            Some(last_seen) => last_seen.saturating_elapsed() > constants::MAX_PEER_TIME_UNSEEN,
+            None => true,
+        }
+    }
```
Comment (teor2345) [inline zebra-network/src/meta_addr.rs]: I like the way these checks are written - they make sure we try each peer at least once.

### Edge Case

What do we want to happen if gossiped `untrusted_last_seen` times are a long way behind our local clock?

With this code, we'll try each peer once, and if it fails, we'll ignore it forever. That seems ok for now. But let's document it.

We might also want to review this design as part of #1873.

### Context

Outdated peer times can happen in at least four different ways:
* the user accidentally sets their clock to the next month or year
* a few peers have their clocks set to the previous month or year
* a peer implementation doesn't provide last seen times (and sets the field to zero)
* malicious peers deliberately gossip lots of peers with very old last seen times

File: zebra-network/src/address_book.rs [hunk 2]
```diff
@@ -304,6 +304,8 @@ impl AddressBook {
         self.by_addr
             .values()
             .filter(|peer| peer.is_ready_for_connection_attempt())
+            // SECURITY: do not keep trying to reconnect to peers that are likely unreachable.
+            .filter(|peer| !peer.is_probably_unreachable())
```
Comment (teor2345) [inline zebra-network/src/address_book.rs]: Naming: It's harder to read code that uses double-negatives. And it's easy to miss the initial `!` operator.

We'd need to invert the logic in `is_probably_reachable` as well.

```rust
            .filter(|peer| peer.is_probably_reachable())
```

File: zebra-network/src/address_book/tests/prop.rs [hunk 3]
```diff
@@ -33,4 +33,19 @@ proptest! {
             prop_assert!(duration_since_last_seen <= MAX_PEER_ACTIVE_FOR_GOSSIP);
         }
     }
+
+    /// Test if peers that are probably unreachable are not listed for a reconnection attempt.
+    #[test]
+    fn probably_unreachable_addresses_are_not_attempted(
```
Comment (teor2345) [inline zebra-network/src/address_book/tests/prop.rs]: Nitpick - Naming

```suggestion
    fn only_reachable_addresses_are_attempted(
```

File: zebra-network/src/constants.rs [hunk 4]
```diff
@@ -78,6 +78,15 @@ pub const MIN_PEER_RECONNECTION_DELAY: Duration = Duration::from_secs(60 + 20 +
...
 pub const MAX_PEER_ACTIVE_FOR_GOSSIP: Duration32 = Duration32::from_hours(3);
 
+/// The maximum duration since a peer was last seen to consider reconnecting to it.
+///
+/// Peers that haven't been seen for more than three days and that had its last connection attempt
+/// fail are considered to be offline and Zebra will stop trying to connect to them.
+///
+/// This is to ensure that Zebra can't have a denial-of-service as a consequence of having too many
+/// offline peers that it constantly and uselessly retries to connect to.
+pub const MAX_PEER_TIME_UNSEEN: Duration32 = Duration32::from_days(3);
```
Comment (teor2345) [inline zebra-network/src/constants.rs]: Naming: It might be easier to understand the code if we say `AGE` rather than `TIME_UNSEEN`.

```suggestion
pub const MAX_RECENT_PEER_AGE: Duration32 = Duration32::from_days(3);
```

File: zebra-network/src/address_book.rs [hunk 5]
```diff
@@ -304,6 +304,8 @@ impl AddressBook {
         self.by_addr
             .values()
             .filter(|peer| peer.is_ready_for_connection_attempt())
```
Comment (teor2345) [inline zebra-network/src/address_book.rs]: Bug Risk: Should we make this new check part of `is_ready_for_connection_attempt`, so it's used everywhere we check for connection attempts?

Or is `is_ready_for_connection_attempt` used for different purposes elsewhere?

File: zebra-network/src/meta_addr.rs [hunk 6]
```diff
@@ -521,6 +521,24 @@ impl MetaAddr {
...
+    /// A peer is considered unreachable if the last connection attempt to it failed and the last
+    /// successful connection is more than 3 days ago.
+    pub fn is_probably_unreachable(&self) -> bool {
+        self.last_connection_state == PeerAddrState::Failed && self.was_not_recently_seen()
+    }
+
+    /// Was this peer last seen a long time ago?
+    ///
+    /// A long time in this case is more than 3 days ago or never seen before.
+    pub fn was_not_recently_seen(&self) -> bool {
```
Comment (teor2345) [inline zebra-network/src/meta_addr.rs]: Naming: Negatives are harder to read and understand.

We'd need to invert the logic in `is_probably_unreachable` as well.

```rust
    pub fn last_seen_is_recent(&self) -> bool {
```

**Review by teor2345 — COMMENTED** [review]
This looks good, I just want to double-check an edge case.

I also have some name suggestions that might make the code easier to read.

(Sorry, I merged the latest main branch into this PR, must have got confused with another one.)

File: zebra-network/src/address_book.rs [hunk 7]
```diff
@@ -304,6 +304,8 @@ impl AddressBook {
         self.by_addr
```
Comment (teor2345) [inline zebra-network/src/address_book.rs]: Usability: Do we need to warn users if they have no recent peers?
Are our existing warnings enough here?

**Review by teor2345 — COMMENTED** [review]
Do we need to test this code in an integration test?
What happens if Zebra has no recent peers?

File: zebra-network/src/meta_addr.rs (hunk 1 above)
Comment (jvff) [reply zebra-network/src/meta_addr.rs]: Done in [01d63d1](https://github.com/ZcashFoundation/zebra/pull/3030/commits/01d63d109500baa7056c8200983c78de4749dd3e). At least I think. Let me know if there's anything else I should do.

File: zebra-network/src/address_book.rs (hunk 2 above)
Comment (jvff) [reply zebra-network/src/address_book.rs]: Done in [b533941](https://github.com/ZcashFoundation/zebra/pull/3030/commits/b533941f80fbae403d2e7d3cf6cd7b2c6f231406).

File: zebra-network/src/address_book.rs (hunk 5 above)
Comment (jvff) [reply zebra-network/src/address_book.rs]: Done in [01d63d1](https://github.com/ZcashFoundation/zebra/pull/3030/commits/01d63d109500baa7056c8200983c78de4749dd3e). Yeah, I think it makes sense to move where the method is called. I also updated the documentation to describe the security implications a bit.

I checked for other usages, and I only found [`AddressBook::maybe_connected_peers`](https://github.com/ZcashFoundation/zebra/blob/main/zebra-network/src/address_book.rs#L324-L331). I think that method might need to be improved in the future.

File: zebra-network/src/meta_addr.rs (hunk 6 above)
Comment (jvff) [reply zebra-network/src/meta_addr.rs]: Done in [c27243a](https://github.com/ZcashFoundation/zebra/pull/3030/commits/c27243ab655028d2fd325bf70970291908140742).

File: zebra-network/src/address_book.rs (hunk 7 above)
Comment (jvff) [reply zebra-network/src/address_book.rs]: I think this might be something for a separate issue?

**Comment by jvff** [comment]
> Do we need to test this code in an integration test? What happens if Zebra has no recent peers?

What do you mean by an integration test? Something like the acceptance tests we have?

File: zebra-network/src/meta_addr.rs (hunk 1 above)
Comment (teor2345) [reply zebra-network/src/meta_addr.rs]: I updated ticket #1873 with this edge case, I think we can deal with any remaining issues there.

File: zebra-network/src/address_book.rs (hunk 5 above)
Comment (teor2345) [reply zebra-network/src/address_book.rs]: The output of `AddressBook::maybe_connected_peers` won't change, because `Failed` peers are never connected.

Also, the maybe connected time is a few minutes, but the `is_probably_reachable` time is days, so the times never overlap either.

File: zebra-network/src/address_book.rs (hunk 7 above)
Comment (teor2345) [reply zebra-network/src/address_book.rs]: Since the existing logs and metrics use `is_ready_for_connection_attempt`, most of them will be handled by the changes in commit 01d63d109500baa7056c8200983c78de4749dd3e.

I also added this to ticket #1873, because we'll have a better idea what we need to log after we've implemented peer deletion.

File: zebra-network/src/address_book/tests/prop.rs [hunk 8]
```diff
@@ -33,4 +33,19 @@ proptest! {
             prop_assert!(duration_since_last_seen <= MAX_PEER_ACTIVE_FOR_GOSSIP);
         }
     }
+
+    /// Test if peers that are probably unreachable are not listed for a reconnection attempt.
```
Comment (teor2345) [inline zebra-network/src/address_book/tests/prop.rs]: Nit: remove negation from the docs as well

```suggestion
    /// Test that only peers that are reachable are listed for reconnection attempts.
```

File: zebra-network/src/meta_addr.rs [hunk 9]
```diff
@@ -530,6 +531,36 @@ impl MetaAddr {
         is_node && self.address_is_valid_for_outbound()
     }
 
+    /// Should this peer considered reachable?
+    ///
+    /// A peer is considered unreachable if the last connection attempt to it failed and the last
+    /// successful connection is more than 3 days ago.
```
Comment (teor2345) [inline zebra-network/src/meta_addr.rs]: Nit: remove negation from the docs as well

I'm not sure if we should mention the `AttemptPending` state here, because it only lasts for a short time (4 seconds), and the `was_connection_recently_attempted` skips it.

```suggestion
    /// A peer is probably reachable if:
    /// - it has never been attempted, or
    /// - the last connection attempt was successful, or
    /// - the last successful connection was less than 3 days ago.
```

**Review by teor2345 — COMMENTED** [review]
Some optional comment tweaks.

**Comment by teor2345** [comment]
> > Do we need to test this code in an integration test? What happens if Zebra has no recent peers?
> 
> What do you mean by an integration test? Something like the acceptance tests we have?

I was thinking of an integration test for all of `zebra_network::init`. I don't think we need to test `zebrad`, because it doesn't modify the address book.

Here's a similar test:
https://github.com/ZcashFoundation/zebra/blob/62bfa15e96740796ecb8d570c37aabaf32017183/zebra-network/src/peer_set/initialize/tests/vectors.rs#L1217-L1229

But you'd want to stuff the active address book with outdated peers.

Here's one way to generate fake peers:
https://github.com/ZcashFoundation/zebra/blob/62bfa15e96740796ecb8d570c37aabaf32017183/zebra-network/src/peer_set/initialize/tests/vectors.rs#L1186-L1194

**Comment by teor2345** [comment]
> I was thinking of an integration test for all of `zebra_network::init`. I don't think we need to test `zebrad`, because it doesn't modify the address book.

Thinking about this a bit more...

The tricky part of this test is working out what the candidate set is doing, or if the network stack has hung. So maybe we'll need to test the crawler task separately?

There are tests for the crawler task in the same module.

**Comment by jvff** [comment]
> The tricky part of this test is working out what the candidate set is doing, or if the network stack has hung. So maybe we'll need to test the crawler task separately?

This seems more complicated than anticipated. Maybe we should do this in a separate PR? I opened #3048 to do it later.

File: zebra-network/src/meta_addr.rs [hunk 10]
```diff
@@ -530,6 +531,38 @@ impl MetaAddr {
...
+    /// [`PeerAddrState::Failed`] before an attempt is made), and if that attempt fails it won't
+    /// try to connect ever again.
+    pub fn is_probably_reachable(&self) -> bool {
+        self.last_connection_state != PeerAddrState::Failed || self.last_seen_is_recent()
+    }
+
+    /// Was this peer last seen recently?
+    ///
+    /// Returns `true` if this peer was last seen at most
+    /// [`MAX_RECENT_PEER_AGE`][constants::MAX_RECENT_PEER_AGE] ago.
```
Comment (teor2345) [inline zebra-network/src/meta_addr.rs:555]:
```suggestion
    /// Returns `true` if this peer was last seen at most
    /// [`MAX_RECENT_PEER_AGE`][constants::MAX_RECENT_PEER_AGE] ago.
    /// Returns false if the peer is outdated, or it has no last seen time.
```

File: zebra-network/src/meta_addr.rs [hunk 11]
```diff
@@ -530,6 +531,38 @@ impl MetaAddr {
...
+    /// # Security
+    ///
+    /// This is used by [`Self::is_ready_for_connection_attempt`] so that Zebra stops trying to
+    /// connect to peers that are likely unreachable.
+    ///
+    /// The `untrusted_last_seen` time is used as a fallback time if the local node has never
+    /// itself seen the peer. If the reported last seen time is a long time ago, then the local
+    /// node will attempt to connect the peer once (because the peer state won't be
+    /// [`PeerAddrState::Failed`] before an attempt is made), and if that attempt fails it won't
+    /// try to connect ever again.
```
Comment (teor2345) [inline zebra-network/src/meta_addr.rs]:
```suggestion
    /// The `untrusted_last_seen` time is used as a fallback time if the local node has never
    /// itself seen the peer. If the reported last seen time is a long time ago or `None`, then the local
    /// node will attempt to connect the peer once, and if that attempt fails it won't
    /// try to connect ever again. (The state can't be `Failed` until after the first connection attempt.)
```

**Comment by teor2345** [comment]
> This seems more complicated than anticipated. Maybe we should do this in a separate PR? I opened #3048 to do it later.

Fair enough. I don't think these code changes are particularly risky by themselves, and we do have decent test coverage of the crawler from other tests in that same module.

I added two specific scenarios that I am concerned about to ticket #3048.

I'll try to be clearer about specific test scenarios next time. Even if we decide to do them in a later ticket, we need to be specific about what we're testing. Otherwise we won't know how to estimate or schedule the ticket.

**Review by teor2345 — COMMENTED** [review]
Let's document what happens when peers have no last seen time.

**Comment by teor2345** [comment]
I tried a manual merge, let's see if it works.

**Review by teor2345 — DISMISSED** [review]

File: zebra-network/src/meta_addr/tests/prop.rs [hunk 12]
```diff
@@ -15,8 +15,10 @@ use tokio::{runtime, time::Instant};
 use tower::service_fn;
 use tracing::Span;
 
+use zebra_chain::serialization::{DateTime32, ZcashDeserialize, ZcashSerialize};
```
Comment (teor2345) [inline zebra-network/src/meta_addr/tests/prop.rs]: Oops, should have used a compiler:
```suggestion
use zebra_chain::serialization::DateTime32;
```

**Review by teor2345 — APPROVED** [review]
